### ISSUE TITLE
S3900: Update documentation to mention protected methods

### ISSUE THREAD

**Issue by elpht** [SonarSource/sonar-dotnet#2336]
### Description

S3900 checks for arguments of protected methods while rule's description states that only public methods are checked.

### Repro steps

Given these two classes pasted below, S3900 complains about CustomRegistry.RegisterServices(Registry registry), last line.

```
    public abstract class RegistryObject
    {
        private Registry structureMapRegistry;

        public Registry StructureMapRegistry => structureMapRegistry ?? (structureMapRegistry = BuildRegistry());

        private Registry BuildRegistry()
        {
            var registry = new Registry();
            registry.For<IService1>().Use<Service1Impl>();
            RegisterServices(registry);
            return registry;
        }

        protected virtual void RegisterServices(Registry registry)
        {
        }
    }

    public class CustomRegistry : RegistryObject
    {
        protected override void RegisterServices(Registry registry)
        {
            base.RegisterServices(registry);
            registry.For<IService2>().Use<Service2Impl>();
        }
    }
```
### Expected behavior

I would expect that the rule doesn't detect anything as the pointed method is "protected".

### Actual behavior

A code smell gets detected.

### Known workarounds

Nothing except disabling the rule.

### Related information

SonarQube 7.6.0.21501

**Comment by Corniel** [comment]
**Arguments of public methods should be validated against null**
https://rules.sonarsource.com/csharp/RSPEC-3900

The reasoning behind this rule is - of course - that you want to be prevent null reference exceptions from (external) calls, that you can not influence. Only with private and internal you have that control, as also protected methods can be implemented by others in other assemblies that forget to not pass `null`s in.

So to me, this warning is valid. You could argue that the documentation should be mentioning that protected methods are also taken into account.

**Comment by elpht** [comment]
Nothing to say about the intention of the rule: my team just activated a bunch of new rules and usually I don't have time to think over the reasoning behind every rule, I just read the documentation.
Just want to notice that if the description did not say "public" or the documentation held your explanation, @Corniel , I would not have open a ticket/issue.
Having said that, thanks for the explanation!

**Comment by Evangelink** [comment]
Hi @elpht and @Corniel.

Thank you for the feedback! I am taking care to update the rule definition to mention `publicly accessible` methods!

**Comment by Evangelink** [comment]
And just to give a bit more context, we are checking the effective accessibility which means that a `public` method inside a `private` class won't be detected by this rule.
